### PR TITLE
v1: Added OnHotkey().

### PR DESCRIPTION
**Pull request by jeeswg**
## Introduction

This function would allow you to execute a callback function, each time a hotkey or hotstring was triggered.
It handles hotkeys, non-auto-replace hotstrings and auto-replace hotstrings.

The callback function receives 4 values:
1. Label. The hotkey/hotstring label.
2. HwndActive. The HWND of the active window when the hotkey was triggered.
3. HwndMatch. The HWND of the window that matched the `#IfWin` criteria, if any. (Note: this may sometimes give useful though unintuitive results, e.g. it returns 0 for a global hotkey/hotstring and it returns 0 for `IfWinNot` criteria.)
4. Type. The hotkey type: 0 if a hotkey, 1 if a non-auto-replace hotstring, 2 if an auto-replace hotstring.

This function or similar would meet the demand for logging hotkey usage data or taking other actions, each time a hotkey/hotstring is triggered.

## Rationale

This functionality is something I'd thought about almost immediately, when I first started using AutoHotkey, around 10 years ago. To log usage data. E.g. first time, last time, count.

Possible workarounds:
- Replace all instances of `return` with a function that logs hotkey usage data and then calls `Exit`. E.g. I use `RetX()` in key scripts.
- Replace all hotkey subroutines with functions, and use the `Hotkey` command/function to launch a go-between function, that logs usage data, before launching the functions.
- (These both have the drawback of not working with auto-replace hotstrings.)

A separate `OnHotstring` function would be fine. Either way, the `Type` parameter allows users to filter by type.

I ordered the type values so that `Type` was effectively a bool, 'IsHotstring': 0 for hotkey, 1 or 2 for hotstring. Hotkeys and non-auto-replace hotstrings are quite similar, so I gave them closer numbers: 0 and 1, leaving auto-replace hotstrings as 2.

I'm not in any great hurry to have this function or similar approved, but I present it here now, if it's of interest.

## Refactoring and 'A_ThisHotkeyHwnd'

tl;dr: IIRC the source code retrieves the active window twice, it could be simpler and more useful/intuitive.

Something I thought of since writing the function: the source code may need some minor refactoring to guarantee 'the active window when the hotkey was triggered'. E.g. the active window HWND could **always** be retrieved, whenever a hotkey/hotstring is triggered, and that HWND would be passed **into**, not out of, any HotCriterionAllowsFiring/CriterionAllowsFiring checks. The 'match' HWND would be unchanged.

I had thought that an `A_ThisHotkeyHwnd` variable might be useful, equivalent to `HwndActive` above.
I would use it in virtually every hotkey subroutine I have. It would be more secure than using `WinExist` to retrieve the HWND retrospectively, which is what I'm doing now.

The point about HotCriterionAllowsFiring and CriterionAllowsFiring would also apply.

An `A_ThisHotkeyHwndMatch` variable might be useful for hotkeys with `#IfWinExist`, however, I don't recall ever needing that, and it could also be achieved by using `#If WinExist(...)` and retrieving the Last Found Window. So I'm not recommending that.

## Test code

```
;==================================================

;test OnHotkey:

OnHotkey("MyOnHotkeyCallbackFunc")
OnHotkey("MyOnHotkeyCallbackFunc2")
return

#IfWinActive ahk_class Notepad
:*b0:myhg2-::my hotstring

#IfWinActive ahk_class Notepad
q::
:*b0:myhg-::
return

#IfWinNotActive ahk_class Notepad
+q::
return

#IfWinExist ahk_class WordPadClass
w::
return

#IfWinNotExist ahk_class WordPadClass
+w::
return

#If WinExist("ahk_class WordPadClass")
r::
return

#If !WinExist("ahk_class WordPadClass")
+r::
return

MyOnHotkeyCallbackFunc(oParams*)
{
	WinGetTitle, vWinTitle1, % "ahk_id " oParams[2]
	WinGetTitle, vWinTitle2, % "ahk_id " oParams[3]
	ToolTip, % oParams[1] "`r`n" vWinTitle1 "`r`n" vWinTitle2 "`r`n" oParams[4]
}

MyOnHotkeyCallbackFunc2(oParams*)
{
	SoundBeep
}

;==================================================
```

### CODE REVIEW

**Comment by krasnovpro** [comment]
Will this feature be added to ahk v2?

**Comment by krasnovpro** [comment]
It would also be nice to get a hotif condition of hotkey.
In these cases:
```
#HotIf CoordMode("Mouse") or (x := y := 0, MouseGetPos(&x, &y), y < 2)
F1:: MsgBox("Any app\Screen ⭱ edge")

#HotIf aiWinActive("main") and (SubStr(ai.appFullVer, 1, 4) < 25.3)
F1:: MsgBox("Adobe Illustrator\Ver<25.3")

#HotIf aiWinActive("main")
F1:: MsgBox("Adobe Illustrator")
```

It will be:
```
CoordMode("Mouse") or (x := y := 0, MouseGetPos(&x, &y), y < 2)
aiWinActive("main") and (SubStr(ai.appFullVer, 1, 4) < 25.3)
aiWinActive("main")
```

**Comment by Lexikos** [comment]
I still don't see the utility of this, and am not interested in considering it further.
